### PR TITLE
Fix MYFACES-4117 in JSF 2.2

### LINKED ISSUE #30414: Port MYFACES-4117 (No default name for @FacesComponent with createTag=true and no tagName)
**Describe the bug**  
No default tag name is provided when creating a custom component with @FacesComponent with createTag=true.

/index.xhtml at line 11 and column 28 <t:someComponent> Tag Library supports namespace: http://xmlns.jcp.org/jsf/component, but no tag was defined for name: someComponent

Stack trace will look similar to: 
```
javax.faces.view.facelets.TagException: /index2.xhtml at line 11 and column 28 <t:dummyComponent> Tag Library supports namespace: http://xmlns.jcp.org/jsf/component, but no tag was defined for name: dummyComponent
	at org.apache.myfaces.view.facelets.compiler.CompilationManager.pushTag(CompilationManager.java:300)
	at org.apache.myfaces.view.facelets.compiler.SAXCompiler$CompilationHandler.startElement(SAXCompiler.java:251)
	at com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.startElement(AbstractSAXParser.java:509)
	at com.sun.org.apache.xerces.internal.parsers.AbstractXMLDocumentParser.emptyElement(AbstractXMLDocumentParser.java:183)
```

**Steps to Reproduce**  
As mentioned, create a component using @FacesComponent with createTag=true.

For example: 

```
@FacesComponent(value = SomeComponent.COMPONENT_TYPE, createTag = true)
public class SomeComponent extends UIComponentBase {
```

**Expected behavior**  
No exceptions should occur.  Developer should be able to use a tag with the default name.  The default name will be the component's class, but the first letter lowercase. The SomeComponent class will use the tag name: `someComponent`.

**Diagnostic information:**  
 - OpenLiberty Version:  All Versions
 - Affected feature(s) jsf-2.2
 - Java Version: N/A

**Additional context**  
https://issues.apache.org/jira/browse/MYFACES-4117
Workaround is use specify a tag name: https://docs.oracle.com/javaee/7/api/javax/faces/component/FacesComponent.html#tagName--

### PR DESCRIPTION
**Pull request by volosied**
- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

fixes #30414

### CODE REVIEW

**Comment by volosied** [comment]
Commit for reference: `https://github.com/apache/myfaces/commit/a5ecbb3657437e79ceb1702d35d6d714740ec213#diff-bbd193714a7a5f6f8534e2762eb571611c2fd671d68838f136247247b5062a02R141`

**Comment by volosied** [comment]
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=901bb70d-2787-43a4-b322-8fe45c95ef03) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

**Comment by volosied** [comment]
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=19d379b3-f802-4cf8-90b9-adb651bdd2fd) - :white_check_mark: completed successfully!)

_Note: Target locations of links might be accessible only to IBM employees._

File: dev/com.ibm.ws.jsf.2.2_fat/fat/src/com/ibm/ws/jsf22/fat/tests/JSF22ComponentRendererTests.java [hunk 1]
```diff
@@ -288,4 +289,28 @@ public void testJsf599() throws Exception {
             assertTrue(page.asText().contains("I'm a place!"));
         }
     }
+
+        /*
+     *  Two items are tested. 
+     *  1) Default tag name is assigned when one is not specified on the @FacesComonent annotation.
+     *  2) Verifies a customer renderer is used when specified. This was added because we don't have any
```
Comment (pnicolucci) [inline dev/com.ibm.ws.jsf.2.2_fat/fat/src/com/ibm/ws/jsf22/fat/tests/JSF22ComponentRendererTests.java]: `custom` or `customer`?

File: dev/com.ibm.ws.jsf.2.2_fat/fat/src/com/ibm/ws/jsf22/fat/tests/JSF22ComponentRendererTests.java [hunk 2]
```diff
@@ -288,4 +289,28 @@ public void testJsf599() throws Exception {
             assertTrue(page.asText().contains("I'm a place!"));
         }
     }
+
+        /*
+     *  Two items are tested. 
+     *  1) Default tag name is assigned when one is not specified on the @FacesComonent annotation.
+     *  2) Verifies a customer renderer is used when specified. This was added because we don't have any 
+     *  exisiting tests for the @FacesRenderer annotation.   The component doesn't render the element, but
+     *  the renderer.
```
Comment (pnicolucci) [inline dev/com.ibm.ws.jsf.2.2_fat/fat/src/com/ibm/ws/jsf22/fat/tests/JSF22ComponentRendererTests.java]: `but the renderer` does what?

File: dev/com.ibm.ws.jsf.2.2_fat/test-applications/JSF22ComponentRenderer.war/resources/myfaces4117.xhtml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+ * Copyright (c) 2015, 2019 IBM Corporation and others.
```
Comment (pnicolucci) [inline dev/com.ibm.ws.jsf.2.2_fat/test-applications/JSF22ComponentRenderer.war/resources/myfaces4117.xhtml]: Is this copyright correct?

File: dev/com.ibm.ws.jsf.2.2_fat/test-applications/JSF22ComponentRenderer.war/resources/myfaces4117.xhtml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License 2.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-2.0/
+ * 
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *     IBM Corporation - initial API and implementation
```
Comment (pnicolucci) [inline dev/com.ibm.ws.jsf.2.2_fat/test-applications/JSF22ComponentRenderer.war/resources/myfaces4117.xhtml]: remove contributors

File: dev/com.ibm.ws.jsf.2.2_fat/test-applications/JSF22ComponentRenderer.war/src/com/ibm/ws/jsf22/fat/componentrenderer/myfaces4117/TestFacesComponent.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    public static final String COMPONENT_TYPE = "custom.type";
+
+    public TestFacesComponent() {
+        setRendererType("test.renderer");
+    }
+    @Override
+    public String getFamily() {        
+        return COMPONENT_FAMILY;
+    }
+}
```
Comment (pnicolucci) [inline dev/com.ibm.ws.jsf.2.2_fat/test-applications/JSF22ComponentRenderer.war/src/com/ibm/ws/jsf22/fat/componentrenderer/myfaces4117/TestFacesComponent.java]: new line

File: dev/com.ibm.ws.jsf.2.2_fat/test-applications/JSF22ComponentRenderer.war/src/com/ibm/ws/jsf22/fat/componentrenderer/myfaces4117/TestFacesRenderer.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+import javax.faces.component.UIComponent;
+import javax.faces.component.UIData;
+import javax.faces.context.FacesContext;
+import javax.faces.context.ResponseWriter;
+import javax.faces.render.FacesRenderer;
+import javax.faces.render.Renderer;
+
+/*
+ * No testing FATS use @FacesRenderer, so this is added to verify the annotation works.
```
Comment (pnicolucci) [inline dev/com.ibm.ws.jsf.2.2_fat/test-applications/JSF22ComponentRenderer.war/src/com/ibm/ws/jsf22/fat/componentrenderer/myfaces4117/TestFacesRenderer.java]: `No existing FATS`

File: dev/com.ibm.ws.jsf.2.2_fat/test-applications/JSF22ComponentRenderer.war/src/com/ibm/ws/jsf22/fat/componentrenderer/myfaces4117/TestFacesRenderer.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+	@Override
+	public void encodeEnd(FacesContext context, UIComponent component) throws IOException {
+		System.out.println("Within TestFacesRenderer#encodeEnd");
+		ResponseWriter writer = context.getResponseWriter();
+        writer.endElement("div");
+		super.encodeEnd(context, component);
+	}
+
+}
```
Comment (pnicolucci) [inline dev/com.ibm.ws.jsf.2.2_fat/test-applications/JSF22ComponentRenderer.war/src/com/ibm/ws/jsf22/fat/componentrenderer/myfaces4117/TestFacesRenderer.java]: new line

File: dev/com.ibm.ws.jsf.2.2_fat/fat/src/com/ibm/ws/jsf22/fat/tests/JSF22ComponentRendererTests.java (hunk 2 above)
Comment (volosied) [reply dev/com.ibm.ws.jsf.2.2_fat/fat/src/com/ibm/ws/jsf22/fat/tests/JSF22ComponentRendererTests.java]: Modified to say: ` The component class doesn't render the element, but
     *  the renderer class does. `

**Review by pnicolucci — APPROVED** [review]

**Comment by volosied** [comment]
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=bcefd486-994d-4f2d-9366-648d4066acbf) - :x: completed with errors/failures.)

_Note: Target locations of links might be accessible only to IBM employees._

**Comment by volosied** [comment]
!build ([view Open Liberty Personal Build](https://libh-proxy1.fyre.ibm.com/cognitive/pipelineAnalysis.html?pipelineId=26602749-5690-42f8-b2ab-27d19fd94aa3) - :white_check_mark: completed successfully!)

_Note: Target locations of links might be accessible only to IBM employees._

**Pull request merged**